speculator/train_speculator.py

```diff
@@ -0,0 +1,321 @@
+        prompt = template.format("Write a bubble sort function in python.")
+
+    tokens = tokenizer.tokenize(prompt)
+    ids = tokenizer.convert_tokens_to_ids(tokens)
+    if "llama" in arch:
+        ids = [tokenizer.bos_token_id] + ids
+    ids = torch.tensor(ids, dtype=torch.long, device="cuda")
+    result = generation.generate(
+        model,
+        ids,
+        max_new_tokens=100,
+        use_cache=True,
+        do_sample=False,
+        max_seq_len=8192,
+    )
+    result = generation.truncate_after_eos(result, tokenizer.eos_token_id)
+    if rank == 0:
+        print(f"{rank}: quick test of base model")
+        print(
+            tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(result))
+        )
+
+
+def main(**kwargs):
+    # get configs
+    cfg = config.train_config()
+    update_config(cfg, **kwargs)
+    cfg.seq_length = cfg.seq_length + cfg.n_speculator_heads + 1
+
+    # ensure reproducibility
+    torch.cuda.manual_seed(cfg.seed)
+    torch.manual_seed(cfg.seed)
+
+    # torchrun specific
+    local_rank = int(os.environ["LOCAL_RANK"])
+    rank = int(os.environ["RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+
+    if rank == 0:
+        print(f"{time.time()} running with these configs {cfg}")
+
+    # some setups
+    torch.cuda.set_device(local_rank)
+
+    if cfg.sharding_strategy != "tp":
+        setup()
+        torch._C._distributed_c10d._register_process_group("default", dist.group.WORLD)
+        base_model_mesh = None
+        speculator_mesh = None
+    else:
+        tp_size = 8
+        base_model_mesh = dist.device_mesh.init_device_mesh(
+            "cuda", (world_size // tp_size, tp_size), mesh_dim_names=("dp", "tp")
+        )
+        speculator_mesh = dist.device_mesh.init_device_mesh("cuda", (world_size,))
+        torch._C._distributed_c10d._register_process_group(
+            "default", base_model_mesh["tp"].get_group()
+        )
+
+    torch.cuda.empty_cache()
+    setup_environ_flags()
+    torch.set_default_dtype(torch.bfloat16)
+
+    # get policy
+    (
+        mixed_precision_policy,
+        wrapping_policy,
+        sharding_strategy_policy,
+        apply_selective_ac,
+        param_init_fn,
+    ) = get_policies(cfg, rank, LLaMABlock)
+
+    model = get_model(
+        cfg.model_arch,
+        cfg.model_variant,
+        model_path=cfg.model_path,
+        device_type="cuda",
+        source="hf",
+        distributed_strategy=cfg.sharding_strategy,
+        group=(
+            base_model_mesh["tp"].get_group() if cfg.sharding_strategy == "tp" else None
+        ),
+    )
+
+    if rank == 0:
+        print(f"{time.time()}")
+        print(model.config)
+        print(model)
+
+    model.eval()
+    with torch.no_grad():
+        test_model(rank, model, cfg.model_arch, cfg, prompt_type="chat")
```

Comment:
can we make prompt_type configurable through the speculator training config?

Reply:
Maybe we could have a test json that can be used (by default not required) and the json would include prompt, expected output, etc. Then we could just add to the config test_metadata param